### ISSUE TITLE
Table as column?

### ISSUE THREAD

**Issue by maqroll** [ClickHouse/ClickHouse#8096]
```
SELECT *
FROM y2018.dist AS t
INNER JOIN tmp.A ON t.a = A.a
INNER JOIN tmp.B ON t.b = B.b
WHERE t.a IN
(
    SELECT a
    FROM tmp.A
)
```
fails with:
Received exception from server (version 19.17.5):
Code: 352. DB::Exception: Received from localhost:9000. DB::Exception: Cannot refer column 'tmp.A' to table.

If table A is in default database (so it doesn't need qualifier) works fine.

A bug?

Version 19.17.5.18

Query doesn't make sense.... simplest example.

[Full script](https://gist.github.com/maqroll/672e40d6dddfa39d6179d763ede85294)

**Comment by filimonov** [comment]
Reproduces on bit simpler case:
```sql
create database y2018;
create database tmp;

create table y2018.vef_shard(a String,b String) ENGINE=MergeTree()
PARTITION BY a ORDER BY b;

create table tmp.A (a String) ENGINE=MergeTree()
PARTITION BY a ORDER BY a;

create table tmp.B (b String) ENGINE=MergeTree()
PARTITION BY b ORDER BY b;

-- Fail!!
select * from y2018.vef_shard t inner join tmp.A on t.a = A.a inner join tmp.B on t.b = B.b WHERE t.a in (select a from tmp.A); 
```

Stacktrace
```
Received exception from server (version 19.18.1):
Code: 352. DB::Exception: Received from localhost:9000. DB::Exception: Cannot refer column 'tmp.A' to table. Stack trace:

0. 0x55ee739d1060 StackTrace::StackTrace() /build/obj-x86_64-linux-gnu/../dbms/src/Common/StackTrace.cpp:228
1. 0x55ee739d0e35 DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int) /build/obj-x86_64-linux-gnu/../dbms/src/Common/Exception.h:27
2. 0x55ee737108a1 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) [clone .cold] /usr/include/c++/9/bits/basic_string.h:222
3. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
4. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
5. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
6. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
7. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
8. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
9. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
10. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
11. 0x55ee76ed91c8 DB::InDepthNodeVisitor<DB::(anonymous namespace)::ColumnAliasesMatcher, true, std::shared_ptr<DB::IAST> const>::visit(std::shared_ptr<DB::IAST> const&) /usr/include/c++/9/bits/stl_iterator.h:829
12. 0x55ee76edabe6 DB::JoinToSubqueryTransformMatcher::visit(DB::ASTSelectQuery&, std::shared_ptr<DB::IAST>&, DB::JoinToSubqueryTransformMatcher::Data&) /usr/include/c++/9/bits/shared_ptr_base.h:729
13. 0x55ee76edccfe DB::JoinToSubqueryTransformMatcher::visit(std::shared_ptr<DB::IAST>&, DB::JoinToSubqueryTransformMatcher::Data&) /build/obj-x86_64-linux-gnu/../dbms/src/Interpreters/JoinToSubqueryTransformVisitor.cpp:377
14. 0x55ee76e38b7d DB::InDepthNodeVisitor<DB::JoinToSubqueryTransformMatcher, true, std::shared_ptr<DB::IAST> >::visit(std::shared_ptr<DB::IAST>&) /usr/include/c++/9/bits/shared_ptr_base.h:1021
15. 0x55ee76e32dcb DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, std::shared_ptr<DB::IBlockInputStream> const&, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) /build/obj-x86_64-linux-gnu/../dbms/src/Interpreters/InterpreterSelectQuery.cpp:259
16. 0x55ee76e339b7 DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, DB::SelectQueryOptions const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) /usr/include/c++/9/bits/shared_ptr_base.h:729
17. 0x55ee76e57bfb DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, DB::Context const&, DB::SelectQueryOptions const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) /usr/include/c++/9/bits/vector.tcc:112
18. 0x55ee76e0f688 DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, DB::Context&, DB::QueryProcessingStage::Enum) /usr/include/c++/9/bits/stl_vector.h:677
19. 0x55ee76f5996a DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*, bool) /build/obj-x86_64-linux-gnu/../dbms/src/Interpreters/executeQuery.cpp:314
20. 0x55ee76f5adee DB::executeQuery(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) /usr/include/c++/9/bits/shared_ptr_base.h:1180
21. 0x55ee73a59b64 DB::TCPHandler::runImpl() /build/obj-x86_64-linux-gnu/../dbms/programs/server/TCPHandler.cpp:247
22. 0x55ee73a5a59b DB::TCPHandler::run() /build/obj-x86_64-linux-gnu/../dbms/programs/server/TCPHandler.cpp:1239
23. 0x55ee77a02fa0 Poco::Net::TCPServerConnection::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:57
24. 0x55ee77a036bd Poco::Net::TCPServerDispatcher::run() /usr/include/c++/9/bits/atomic_base.h:327
25. 0x55ee790d9831 Poco::PooledThread::run() /usr/include/x86_64-linux-gnu/c++/9/bits/gthr-default.h:748
26. 0x55ee790d75dc Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/include/Poco/AutoPtr.h:205
27. 0x55ee7984d740 execute_native_thread_routine /usr/lib/debug/usr/bin/clickhouse
28. 0x7facf5b456db start_thread /lib/x86_64-linux-gnu/libpthread-2.27.so
29. 0x7facf546288f clone /build/glibc-OTsEL5/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```
/cc @4ertus2

**Comment by maqroll** [comment]
Thanks!!

Just a comment:

it works if you take out any of the inner joins.
Either
`
select * from y2018.vef_shard t 
--inner join tmp.A on t.a = A.a 
inner join tmp.B on t.b = B.b 
WHERE t.a in (select a from tmp.A);
`

or

`
select * from y2018.vef_shard t 
inner join tmp.A on t.a = A.a 
--inner join tmp.B on t.b = B.b 
WHERE t.a in (select a from tmp.A);
`

**Comment by inakivb** [comment]
Hello,

I'm having the same problem. Is there any workarround to make two joins and also a IN ?

Thank you!

**Comment by maqroll** [comment]
@inakivb which version are you using?

**Comment by maqroll** [comment]
Seems to be fixed in version 20.3.8.53.
